source/common/conn_pool/conn_pool_base.cc

```diff
@@ -427,6 +458,7 @@ void ConnPoolImplBase::onPendingStreamCancel(PendingStream& stream,
 
   host_->cluster().stats().upstream_rq_cancelled_.inc();
   checkForDrained();
+  checkForIdle();
```

Comment:
I would have to dive back in and fully revisit the logic, but it seems to me that `checkForDrained` and `checkForIdle` are doing very similar things and have to be called at all the same points. Can they be merged into `checkForDrained` to limit the required changes in this PR as well as logic divergence?

Reply:
I considered whether the logic could be merged, but I think there are subtle differences that prevent that (I think).

`checkForDrained` does the following:
* Closes connections in the `READY` state (and `CONNECTING` if there are no pending streams)
* Fires the drained callbacks if there are no pending, ready, busy, and connecting clients

For `checkForIdle`, we don't want to change the state of the connections at all. I think we also primarily care about active and pending _streams_ rather than the connection state (e.g. if the pool has a bunch of `CONNECTING` connections but no pending streams, it is still idle).

There are also a few cases where we need to check for idle but we don't want to check for drained, e.g. when a stream closes we want to check whether we just went idle, but it doesn't make sense to check if we've drained since it's unlikely the connection bearing the closed stream has been drained yet.

Does that make sense?

Reply:
I'm probably missing something, but my general take is that "idle" and "drained" is basically the same thing. We might do different things (fire drained callback or start timer) but I don't quite understand how they are different in that we are checking to see if there are no active connections.

Can we unify into a single function that takes an enum param if that matters?

My general take is it's going to be pretty hard for the next reader to understand how these things are different and we should try to unify them. cc @alyssawilk on this point also.

Reply:
I added some comments to distinguish between these cases -- I hope it's clearer now.

When a pool is "idle" it means that there may be live connections, but those connections do not have any active streams. If the pool sees activity, it may cease to become idle. When the idle timer fires, the pool is drained and removed from the map.

Prior to this PR, a pool is only put into the "draining" state under (afaik) two conditions: the host is unhealthy or the host is being removed. This adds a third state -- the pool has been idle for the idle timeout duration.